Rework custom schema docs (#4815)
fishtown-analytics/docs.getdbt.com: website/docs/docs/build/custom-schemas.md

pagination_next: "docs/build/custom-databases"
---

> By default, all dbt models are built in the schema specified in your [environment](/docs/dbt-cloud-environments) (dbt Cloud) or [profile's target](/docs/core/dbt-core-environments) (dbt Core). This default schema is called your **target schema**.

Reviewer: sugg: per style guide, use italics for emphasis not bold

```suggestion
By default, all dbt models are built in the schema specified in your [environment](/docs/dbt-cloud-environments) (dbt Cloud) or [profile's target](/docs/core/dbt-core-environments) (dbt Core). This default schema is called your _target schema_.
```

> In dbt projects with lots of models, it is often preferable to build models across multiple schemas and group similar models together. For example, you may wish to:

Reviewer: ```suggestion
For dbt projects with lots of models, it's common to build models across multiple schemas and group similar models together. For example, you might want to:
```

> * Group models based on the business unit using the model, creating schemas such as `core`, `marketing`, `finance` and `support`; or,

Reviewer: ```suggestion
* Group models based on the business unit using the model, creating schemas such as `core`, `marketing`, `finance` and `support`
```

* Hide intermediate models in a `staging` schema, and only present models that should be queried by an end user in an `analytics` schema.

> To do this, specify a custom schema. dbt will then generate the schema name for a model by **appending the custom schema to the target schema**, as in: `<target_schema>_<custom_schema>`.

Reviewer: ```suggestion
To do this, specify a custom schema. dbt generates the schema name for a model by appending the custom schema to the target schema. For example, `<target_schema>_<custom_schema>`.
```

| Target schema | Custom schema | Resulting schema |
| ------------- | ------------- | ---------------- |
| analytics_prod | None | analytics_prod |
| alice_dev | None | alice_dev |
| dbt_cloud_pr_123_456 | None | dbt_cloud_pr_123_456 |
| analytics_prod | marketing | analytics_prod_marketing |
| alice_dev | marketing | alice_dev_marketing |
| dbt_cloud_pr_123_456 | marketing | dbt_cloud_pr_123_456_marketing |

## How do I use custom schemas?

> Use the `schema` configuration key. As with any configuration, you can either:

Reviewer: ```suggestion
To specify a custom schema for a model, use the `schema` configuration key. As with any configuration, you can do one of the following:
```

> * apply this configuration to a specific model by using a config block within a model, or

Reviewer: ```suggestion
* apply this configuration to a specific model by using a config block within a model
```

pagination_next: "docs/build/custom-databases"
---

By default, all dbt models are built in the schema specified in your [environment](/docs/dbt-cloud-environments) (dbt Cloud) or [profile's target](/docs/core/dbt-core-environments) (dbt Core). This default schema is called your _target schema_.

For dbt projects with lots of models, it's common to build models across multiple schemas and group similar models together. For example, you might want to:

> * Group models based on the business unit using the model, creating schemas such as `core`, `marketing`, `finance` and `support`

Reviewer: sugg: for parallel construction of the bulleted list (end with punctuation) 

```suggestion
* Group models based on the business unit using the model, creating schemas such as `core`, `marketing`, `finance` and `support`.
```

## Understanding custom schemas

> When first using custom schemas, it's a common misunderstanding to assume that a model will use _only_ the new `schema` configuration, for example, a model that has the configuration `schema: marketing` would be built in the `marketing` schema. However, dbt will actually put it in a schema like `<target_schema>_marketing`.

Reviewer: ```suggestion
When first using custom schemas, it's a common misunderstanding to assume that a model _only_ uses the new `schema` configuration; for example, a model that has the configuration `schema: marketing` would be built in the `marketing` schema. However, dbt puts it in a schema like `<target_schema>_marketing`.
```

> There's a good reason for this deviation! Each dbt user has their own target schema for development (see [Managing Environments](#managing-environments)). If dbt ignored the target schema and only used the model's custom schema, every dbt user would create models in the same schema and would overwrite each other's work.

Reviewer: ```suggestion
There's a good reason for this deviation. Each dbt user has their own target schema for development (refer to [Managing Environments](#managing-environments)). If dbt ignored the target schema and only used the model's custom schema, every dbt user would create models in the same schema and would overwrite each other's work.
```

{%- endmacro %}
```

## Changing the way dbt generates a schema name

If your dbt project has a custom macro called `generate_schema_name`, dbt will use it instead of the default macro. This allows you to customize the name generation according to your needs.

> To customize this macro, copy the example code above into a file named `macros/generate_schema_name.sql` and make changes as necessary.

Reviewer: sugg: we've started to avoid using words like "below" and "above" to use more inclusive language for less-abled people

```suggestion
To customize this macro, copy the example code in the section [How does dbt generate a model's schema name](#how-does-dbt-generate-a-models-schema-name) into a file named `macros/generate_schema_name.sql` and make changes as necessary.
```

> **Note**: dbt will ignore any custom `generate_schema_name` macros included in installed packages.

Reviewer: sugg: bold is reserved for UI labels, page titles, etc.

```suggestion
Be careful. dbt will ignore any custom `generate_schema_name` macros included in installed packages.
```

</VersionBlock>

## A built-in alternative pattern for generating schema names

A common customization is to ignore the target schema in production environments, and ignore the custom schema configurations in other environments (such as development and CI).

Production Environment (`target.name == 'prod'`)

| Target schema | Custom schema | Resulting schema |
| ------------- | ------------- | ---------------- |
| analytics_prod | None | analytics_prod |
| analytics_prod | marketing | marketing |

Development/CI Environment (`target.name != 'prod'`)

| Target schema | Custom schema | Resulting schema |
| ------------- | ------------- | ---------------- |
| alice_dev | None | alice_dev |
| alice_dev | marketing | alice_dev |
| dbt_cloud_pr_123_456 | None | dbt_cloud_pr_123_456 |
| dbt_cloud_pr_123_456 | marketing | dbt_cloud_pr_123_456 |

> Just like the normal macro, this approach guarantees that schemas from different environments will not collide.

Reviewer: ```suggestion
Similar to the regular macro, this approach guarantees that schemas from different environments will not collide.
```

> dbt ships with a macro for this use case – called `generate_schema_name_for_env` – which is disabled by default. To enable it, add a custom `generate_schema_name` macro to your project that contains the following code:

Reviewer: sugg: em dashes are preferred over en dashes in user docs

```suggestion
dbt ships with a macro for this use case &mdash; called `generate_schema_name_for_env` &mdash; which is disabled by default. To enable it, add a custom `generate_schema_name` macro to your project that contains the following code:
```

<File name='macros/get_custom_schema.sql'>

```sql
-- put this in macros/get_custom_schema.sql

{% macro generate_schema_name(custom_schema_name, node) -%}
    {{ generate_schema_name_for_env(custom_schema_name, node) }}
{%- endmacro %}
```

</File>

> **Note:** When using this macro, you'll need to set the target name in your production job to `prod`.

Reviewer: ```suggestion
When using this macro, you'll need to set the target name in your production job to `prod`.
```

## Managing environments

> In the `generate_schema_name` macro examples shown above, the `target.name` context variable is used to change the schema name that dbt generates for models. If the `generate_schema_name` macro in your project uses the `target.name` context variable, you must ensure that your different dbt environments are configured accordingly. While you can use any naming scheme you'd like, we typically recommend:

Reviewer: ```suggestion
In the `generate_schema_name` macro examples shown in the section [](/docs/build/custom-schemas#a-built-in-alternative-pattern-for-generating-schema-names), the `target.name` context variable is used to change the schema name that dbt generates for models. If the `generate_schema_name` macro in your project uses the `target.name` context variable, you must ensure that your different dbt environments are configured accordingly. While you can use any naming scheme you'd like, dbt Labs typically recommends:
```

Author: Handled this one separately as your suggestion didn't have anchor text, but it's resolved now

> * **dev**: Your local development environment; configured in a `profiles.yml` file on your computer.

Reviewer: ```suggestion
* **dev** &mdash; Your local development environment; configured in a `profiles.yml` file on your computer.
```

> * **ci**:  A [continuous integration](/docs/cloud/git/connect-github) environment running on Pull Requests in GitHub, GitLab, etc.

Reviewer: ```suggestion
* **ci** &mdash; A [continuous integration](/docs/cloud/git/connect-github) environment running on pull pequests in GitHub, GitLab, and so on.
```

> * **prod**: The production deployment of your dbt project, like in dbt Cloud, Airflow, or [similar](/docs/deploy/deployments).

Reviewer: ```suggestion
* **prod** &mdash; The production deployment of your dbt project, like in dbt Cloud, Airflow, or [similar](/docs/deploy/deployments).
```

> If your schema names are being generated incorrectly, double check your target name in the relevant environment.

Reviewer: ```suggestion
If your schema names are being generated incorrectly, double-check your target name in the relevant environment.
```
